build-tools/packages/build-cli/src/commands/generate/releaseNotes.ts

```diff
@@ -0,0 +1,205 @@
+import admonitions from "remark-github-beta-blockquote-admonitions";
+import remarkToc from "remark-toc";
+
+import { releaseGroupFlag } from "../../flags.js";
+import {
+	BaseCommand,
+	DEFAULT_CHANGESET_PATH,
+	FluidCustomChangeSetMetadataDefaults,
+	UNKNOWN_SECTION,
+	difference,
+	groupBySection,
+	loadChangesets,
+} from "../../library/index.js";
+
+const prLinkPrefix = `https://github.com/microsoft/FluidFramework/pull/`;
+
+/**
+ * Generates a summary of all changesets and outputs the results to a file. This is used to generate an UPCOMING.md file
+ * that provides a single place where developers can see upcoming changes.
+ */
+export default class GenerateReleaseNotesCommand extends BaseCommand<
+	typeof GenerateReleaseNotesCommand
+> {
+	static readonly summary = ``;
+
+	// Enables the global JSON flag in oclif.
+	static readonly enableJsonFlag = true;
+
+	static readonly flags = {
+		releaseGroup: releaseGroupFlag({
+			required: true,
+		}),
+		releaseType: Flags.custom<"major" | "minor">({
+			char: "t",
+			description: "The type of release for which the release notes are being generated.",
+			options: ["major", "minor"],
+			required: true,
+			parse: async (input) => {
+				if (input === "major" || input === "minor") {
+					return input;
+				}
+
+				throw new Error(`Invalid release type: ${input}`);
+			},
+		})(),
+		out: Flags.file({
+			description: `Output the results to this file.`,
+			required: true,
+			default: "RELEASE_NOTES.md",
+			// exists: true
+		}),
+		...BaseCommand.flags,
+	} as const;
+
+	static readonly examples = [
+		{
+			description: `Generate release notes for a minor release of the client release group.`,
+			command: "<%= config.bin %> <%= command.id %> -g client -t minor",
+		},
+		{
+			description: `You can output a different file using the --out flag.`,
+			command:
+				"<%= config.bin %> <%= command.id %> -g client -t minor --out RELEASE_NOTES/2.1.0.md",
+		},
+	];
+
+	public async run(): Promise<string> {
+		const context = await this.getContext();
+		const { flags, logger } = this;
+
+		const releaseGroup = context.repo.releaseGroups.get(flags.releaseGroup);
+		if (releaseGroup === undefined) {
+			this.errorLog(`Unknown release group: ${flags.releaseGroup}`);
+			this.exit(2);
+		}
+
+		const { releaseNotes: releaseNotesConfig } = loadFluidBuildConfig(
+			context.gitRepo.resolvedRoot,
+		);
+		if (releaseNotesConfig === undefined) {
+			this.errorLog(
+				`No releaseNotes config found. Make sure the 'releaseNotes' section of the build config exists.`,
+			);
+			this.exit(2);
+		}
+
+		const changesetDir = path.join(releaseGroup.directory, DEFAULT_CHANGESET_PATH);
+		const changesets = await loadChangesets(changesetDir, logger);
+
+		const { version } = releaseGroup;
+		const header = `<!-- THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY. -->`;
+		const footer = `### 🛠️ Start Building Today!\n\nPlease continue to engage with us on GitHub
```

Comment:
Nit: I wonder if it might be nicer to build something like an mdast tree, and convert that to a markdown string at the end, rather than manually building a string with hard-coded heading levels and such.

Probably not for this PR, but maybe something to consider for the future.

Reply:
Definitely something to consider in the future. This was quicker to iterate on for me, but mdast should absolutely be on the table long-term.